compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2308,3 +2308,10 @@ impl<'tcx> VarianceDiagInfo<'tcx> {
         }
     }
 }
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(TyEncodable, TyDecodable, Debug, HashStable, TypeFoldable)]
+pub struct GeneratorWitnessInner<'tcx> {
+    pub tys: &'tcx List<Ty<'tcx>>,
+    pub structural_predicates: &'tcx List<ty::ProjectionPredicate<'tcx>>,
```

Comment:
Yeah, the gist is that a lot of these "not general enough" errors come from not being able to normalize projection types inside of ADTs after we erase the regions in the generator (because the impl that would solve that projection ty no longer holds with the erased regions). So we stash the projection ty => normalized ty as a ProjectionPredicate and add that to the param env when we're checking auto traits.

For example, given a future like `Map<&_#0r T, fn(&_#1r T) -> S>`, we know that future's `Output = S`, but we are unable to determine that the type even implements `Future` when we erase the region variables `_#0r` and `_#1r` later in another, because we can't relate the erased regions anymore.

So the first commit in this PR introduces `StructuralPredicateElaborator` that walks thru the types in the generator interior and yields these normalized projections as ProjectionPredicates. The second commit tries to generalize this to some extent, storing some of the predicates that we know in our ParamEnv too.